server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala

```diff
@@ -31,42 +40,69 @@ object StpUtil {
     name.tail
   }
 
-  def readPreviousTokens(baseUrl: String, path: String, format: String)(implicit context: ExecutionContext) = {
-
-    import cmwell.util.http.SimpleResponse.Implicits.UTF8StringHandler
-
-    cmwell.util.http.SimpleHttpClient
-      .get(s"http://$baseUrl$path/tokens?op=stream&recursive&format=json")
-      .map{
-        case response if response.status == 200 || response.status == 404 =>
-          Right(response.payload.lines
-            .map({
-              row =>
-                parse(row) match {
-                  case Left(parseFailure @ ParsingFailure(_, _)) => throw parseFailure
-                  case Right(json) => {
-
-                    val token = (for {
-                      vec <- json.hcursor.downField("fields").downField("token").values
-                      jsn <- vec.headOption
-                      str <- jsn.asString
-                    } yield str).getOrElse("")
-
-                    val receivedInfotons: Option[DownloadStats] =
-                      json.hcursor.downField("fields").downField("receivedInfotons").downArray.as[Long].toOption.map {
-                        value =>
-                          DownloadStats(receivedInfotons = value)
-                      }
-
-                    val sensor = extractLastPart(json.hcursor.downField("system").get[String]("path").toOption.get)
-                    sensor -> (token, receivedInfotons)
-                  }
+  def readPreviousTokensWithRetry(baseUrl: String, path: String, format: String, zStore: ZStore)
+                                 (implicit context: ExecutionContext) = {
+
+    def shouldRetry(action: String): (Try[TokenAndStatisticsMap], RetryParams) => ShouldRetry[RetryParams] = {
+      import scala.language.implicitConversions
+      implicit def asFiniteDuration(d: Duration) = scala.concurrent.duration.Duration.fromNanos(d.toNanos);
+      {
+        case (Success(_),_) =>
+          logger.debug(s"Successfully read token and statistics state from zStore for agent ${extractLastPart(path)}")
+          DoNotRetry
+        case (Failure(ex), state) if state.retriesLeft == 0 => {
+          logger.error(ex.getMessage)
+          logger.error(s"Failed to read token and statistics state from zStore. 0 retries left. Will not re-attempt for agent ${extractLastPart(path)}")
```

Comment:
I'd combine these two logger.error lines into one, first the s"Faild ..." string and then `, ex)` (without .getMessage)